### PR TITLE
Module feature inspired by Geb

### PR DESCRIPTION
**Pull request by nielsfalk**

### CODE REVIEW

File: src/test/kotlin/com/github/epadronu/balin/core/PageSpec.kt [hunk 1]
```diff
@@ -28,6 +28,10 @@ import org.openqa.selenium.htmlunit.HtmlUnitDriver
 import com.github.epadronu.balin.exceptions.PageAtValidationError
 import com.github.epadronu.balin.extensions.`$`
 import com.github.epadronu.balin.libs.delegatesTo
+import org.jetbrains.spek.api.dsl.given
+import org.jetbrains.spek.api.dsl.it
+import org.jetbrains.spek.api.dsl.on
+
```
Comment (EPadronU) [inline src/test/kotlin/com/github/epadronu/balin/core/PageSpec.kt:34]: This kind of empty lines should not exist. I know Intellij IDEA introduces them but they should be removed.

File: build.gradle [hunk 2]
```diff
@@ -39,7 +39,10 @@ dependencies {
   compile "org.seleniumhq.selenium:selenium-support:$seleniumVersion"
   compile 'org.slf4j:slf4j-api:1.7.21'
 
-  testCompile "org.jetbrains.spek:spek:$spekVersion"
+  testCompile "org.jetbrains.spek:spek-api:$spekVersion"
+  testCompile "org.jetbrains.spek:spek-junit-platform-engine:$spekVersion"
+  testCompile "org.jetbrains.kotlin:kotlin-test:1.1.1"
```
Comment (EPadronU) [inline build.gradle:44]: Please introduce inside the ext block the version number for kotlin-test.

File: build.gradle [hunk 3]
```diff
@@ -39,7 +39,10 @@ dependencies {
...
+  testCompile 'junit:junit:4.12'
```
Comment (EPadronU) [inline build.gradle:45]: Please introduce inside the ext block the version number for junit.

File: src/test/kotlin/com/github/epadronu/balin/core/BrowserSpec.kt [hunk 4]
```diff
@@ -16,285 +16,306 @@
...
 import org.jetbrains.spek.api.Spek
+import org.jetbrains.spek.api.dsl.given
+import org.jetbrains.spek.api.dsl.it
+import org.jetbrains.spek.api.dsl.on
 import org.openqa.selenium.By
 import org.openqa.selenium.TimeoutException
 import org.openqa.selenium.htmlunit.HtmlUnitDriver
-import org.openqa.selenium.support.ui.ExpectedConditions.*
+import org.openqa.selenium.support.ui.ExpectedConditions.presenceOfElementLocated
+import kotlin.test.assertEquals
```
Comment (EPadronU) [inline src/test/kotlin/com/github/epadronu/balin/core/BrowserSpec.kt:37]: Like mentioned elsewhere, this kind of empty lines should be removed.

File: src/test/kotlin/com/github/epadronu/balin/core/BrowserSpec.kt [hunk 5]
```diff
@@ -16,285 +16,306 @@
 
 /* ***************************************************************************/
 package com.github.epadronu.balin.core
+
 /* ***************************************************************************/
 
 /* ***************************************************************************/
```
Comment (EPadronU) [inline src/test/kotlin/com/github/epadronu/balin/core/BrowserSpec.kt:22]: Use this separators in the files you created. I know it is a personal taste but it makes the different sections of a Kotlin file more explicit and clear.

File: src/main/kotlin/com/github/epadronu/balin/core/Module.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        return baseElement.findElement(by)
+    }
+
+    override fun findElements(by: By?): MutableList<WebElement> {
+        return baseElement.findElements(by)
+    }
+}
+
+fun <M : Module> List<WebElement>.module(factory: (WebElement) -> M): M {
+    return factory(this.first())
```
Comment (EPadronU) [inline src/main/kotlin/com/github/epadronu/balin/core/Module.kt:37]: What happens if the list is empty? I could be wrong but shouldn't an exception will be thrown? I don't know if that's the best we can do on such situation.

**Review by EPadronU — CHANGES_REQUESTED** [review]
This is an awesome work. I really liked what you did and I 'm very thankful for your interest in Balin.

Just a tiny general observation: Try to stick to the two-spaces indentation. I know is quite odd for the Java world but even so, it is the recommended amount in the Google Java Style Guide and is the one I used. Sticking to such style of indentation will avoid a great amount of differences in the BrowserSpek.kt file that are not real code changes. Additionally, it will make the indentation among the project consistent.

**Comment by nielsfalk** [comment]
I'm glad you liked my changes.
I'm ok with any codeformat, as long I can use a Codeformatter. I press cmd+alt+l all the time ;) Can you add the formatter-settings to this repo?

**Comment by EPadronU** [comment]
At the moment, I don't have anything fancy but a two-spaces indentation. Maybe when I have the time to continue with the development I'll adopt the one from https://github.com/google/styleguide.

The other aspects (eg: the blank lines and the section separators) has been managed by me in a manual way.

**Comment by dnguyenminh** [comment]
Hi @nielsfalk, 
I got your code, the code was compiled successfully but the tests were not run. How need I do to force the test to run ?

Hi @EPadronU,
I got your code, I ran the the tests and I found some small issues:
- Kotlin index page are updated so your test is out of date, I fixed the validation.
- The browser test about javascript support failed because the core of http unit initialize.

Thank for your code, your idea is clean so I like your ii very much.

**Comment by EPadronU** [comment]
Hi @dnguyenminh. I took some time to update the dependencies and the tests. Try again with the new changes.

**Comment by EPadronU** [comment]
The code base has changed a lot since this pull request and a similar feature has been provided.

**Pull request closed without merging**